### PR TITLE
[CLI] extension update command

### PR DESCRIPTION
**Pull request by samhanic**
This PR adds a cli command to update all extensions `code --list-extensions` , which works without opening VSCode as GUI.
This fixes the issue https://github.com/microsoft/vscode/issues/56578 .

For reference, this is the following of the ticket re-opening after the discussion in this issue https://github.com/microsoft/vscode/issues/196155#issuecomment-1825617895 .

This have been linted using yarn and tested with the provided test.sh script.
I also tested that with classic extensions, but due to difficulties to test, I have not tested any prerelease upgrades, and not tested with vscode server. For prerelease, as soon as I will have some outdated one in my test setup I will be able to update this comment accordingly.
I am also not a TypeScript developer so if I wrote someting not optimal in the language, don't hesitate to state it below.

### CODE REVIEW

File: src/vs/platform/environment/node/argv.ts [hunk 1]
```diff
@@ -100,6 +100,7 @@ export const OPTIONS: OptionDescriptions<Required<NativeParsedArgs>> = {
 	'install-extension': { type: 'string[]', cat: 'e', args: 'ext-id | path', description: localize('installExtension', "Installs or updates an extension. The argument is either an extension id or a path to a VSIX. The identifier of an extension is '${publisher}.${name}'. Use '--force' argument to update to latest version. To install a specific version provide '@${version}'. For example: 'vscode.csharp@1.2.3'.") },
 	'pre-release': { type: 'boolean', cat: 'e', description: localize('install prerelease', "Installs the pre-release version of the extension, when using --install-extension") },
 	'uninstall-extension': { type: 'string[]', cat: 'e', args: 'ext-id', description: localize('uninstallExtension', "Uninstalls an extension.") },
+	'update-extensions': { type: 'boolean', cat: 'e', description: localize('updateExtensions', "Update the marketplace-installed extensions.") },
```
Comment (sandy081) [inline src/vs/platform/environment/node/argv.ts:103]:
```suggestion
	'update-extensions': { type: 'boolean', cat: 'e', description: localize('updateExtensions', "Update the installed extensions.") },
```

File: src/vs/platform/extensionManagement/common/extensionManagementCLI.ts [hunk 2]
```diff
@@ -172,6 +172,57 @@ export class ExtensionManagementCLI {
...
+		const installedExtensions = await this.extensionManagementService.getInstalled(ExtensionType.User, profileLocation);
+
+		const installedExtensionsQuery: IExtensionInfo[] = [];
+		for (const extension of installedExtensions) {
+			if (!!extension.identifier.uuid) { // No need to check new version for an unpublished extension
+				installedExtensionsQuery.push({
+					id: `${extension.manifest.publisher}.${extension.manifest.name}`,
+					preRelease: extension.isPreReleaseVersion
+				});
+			}
```
Comment (sandy081) [inline src/vs/platform/extensionManagement/common/extensionManagementCLI.ts:185]:
```suggestion
		installedExtensionsQuery.push({ ...extension.identifier, preRelease: extension.preRelease });
```

**Comment by connor4312** [comment]
Rust changes look ok 👍

**Comment by sandy081** [comment]
> I choosed the simplest solution and renamed the one in extensionManagementCLI.ts to CLIInstallExtensionInfo , but this could be the topic of a refactoring.

This is ok for me. Let me know once you have incorporated all changes.

**Comment by samhanic** [comment]
I already incorporated all changes you requested, so if renaming as `CLIInstallExtensionInfo` works for you, from what I see it should be ready to merge

File: src/vs/platform/extensionManagement/common/extensionManagementCLI.ts [hunk 3]
```diff
@@ -172,6 +172,41 @@ export class ExtensionManagementCLI {
...
+						extension: newVersion,
+						options: { operation: InstallOperation.Update, installPreReleaseVersion: oldVersion.isPreReleaseVersion }
+					});
+				}
+			}
+		}
+
+		if (extensionsToUpdate.length) {
+			this.logger.info(localize('updateExtensionsNewVersionsAvailable', "Updating extensions: {0}", extensionsToUpdate.map(ext => ext.extension.identifier.id).join(', ')));
+			await this.extensionManagementService.installGalleryExtensions(extensionsToUpdate);
```
Comment (sandy081) [inline src/vs/platform/extensionManagement/common/extensionManagementCLI.ts:202]: Use the return result of `installGalleryExtensions` to log succuessfully updated and failed extensions

Comment (samhanic) [reply src/vs/platform/extensionManagement/common/extensionManagementCLI.ts]: Ok, now I understand better what you meant by "update the logic accordingly".
Last commit pushed should resolve that. Is it verbose in the right amount?

Successful update log example:
```
Updating extensions: eamodio.gitlens, tamasfe.even-better-toml
Successfully updated 2 extensions
```

Failed update log example:
```
Updating extensions: eamodio.gitlens, tamasfe.even-better-toml
Failed updating extensions: tamasfe.even-better-toml, eamodio.gitlens
```

Log for no update available:
```
No extension to update
```

File: src/vs/platform/extensionManagement/common/extensionManagementCLI.ts [hunk 4]
```diff
@@ -172,6 +172,47 @@ export class ExtensionManagementCLI {
...
+
+		this.logger.info(localize('updateExtensionsNewVersionsAvailable', "Updating extensions: {0}", extensionsToUpdate.map(ext => ext.extension.identifier.id).join(', ')));
+		const installationResult = await this.extensionManagementService.installGalleryExtensions(extensionsToUpdate);
+
+		const failed: string[] = installationResult.filter(ext => ext.error).map(ext => ext.identifier.id);
+		if (failed.length) {
+			throw new Error(localize('updateExtensionsFailed', "Failed updating extensions: {0}", failed.join(', ')));
+		}
+
+		this.logger.info(localize('updateExtensionsDone', "Successfully updated {0} extensions", extensionsToUpdate.length));
```
Comment (sandy081) [inline src/vs/platform/extensionManagement/common/extensionManagementCLI.ts:213]:
```suggestion
	for (const extensionResult of installationResult) {
		if (extensionResult.error) {
			this.logger.error(localize('errorUpdatingExtension', "Error while updating extension {0}: {1}", extensionResult.identifier.id, getErrorMessage(extensionResult.error)));
		} else {
			this.logger.info(localize('successUpdate', "Extension '{0}' v{1} was successfully updated.", extensionResult.identifier.id, extensionResult.local.manifest.version));
		}
	}
```

Comment (samhanic) [reply src/vs/platform/extensionManagement/common/extensionManagementCLI.ts]: Ok, just changed `extensionResult.local.manifest.version` to `extensionResult.local?.manifest.version` while commiting suggestion (probably a typo, avoids the error : `'extensionResult.local' is possibly 'undefined'.ts(18048)` )

**Comment by sandy081** [comment]
LGTM, can you please resolve the merge conflicts?

**Review by sandy081 — APPROVED** [review]

**Review by joaomoreno — APPROVED** [review]

**Comment by sandy081** [comment]
> but prints a DepreciationWarning (probably inside the extensionManagementService.installGalleryExtensions function):

This is being tracked separately. Tagging this as verified based on above comment

**Comment by starball5** [comment]
Does this update _all_ extensions? or only all the extensions for the specified profile (`--profile`) / the default if no profile is specified?

**Comment by sandy081** [comment]
Profile should be passed to update for a specific profile, otherwise extensions from default profile are updated.

**Pull request merged**